protocol/clproto_cpp/test/tests/test_parameter_proto.cpp

```diff
@@ -15,110 +39,30 @@ static void test_parameter_equal(const T& send_state, const T& recv_state) {
-
-  T recv_state("");
-  EXPECT_NO_THROW(recv_state = clproto::decode<T>(msg));
-  EXPECT_TRUE(recv_state.is_empty());
-}
+class ParameterProtoTest : public testing::Test {
+public:
+  ParameterProtoTest() : test_case_{std::get<ParamT<T>>(parameter_test_cases)} {}
+protected:
+  ParamT<T> test_case_;
+};
+TYPED_TEST_SUITE_P(ParameterProtoTest);
 
 TEST(ParameterProtoTest, EncodeParameterInterface) {
   auto send_state_ptr = make_shared_state(ParameterInterface("name", ParameterType::BOOL));
   EXPECT_THROW(std::string msg = clproto::encode(send_state_ptr), std::invalid_argument);
 }
 
-TEST(ParameterProtoTest, EncodeDecodeParameter) {
-  test_encode_decode_parameter(
-      Parameter<int>("A", 1), clproto::MessageType::PARAMETER_MESSAGE, clproto::ParameterMessageType::INT
-  );
-  test_encode_decode_parameter(
-      Parameter<std::vector<int>>("A", {1, 2, 3}), clproto::MessageType::PARAMETER_MESSAGE,
-      clproto::ParameterMessageType::INT_ARRAY
-  );
-  test_encode_decode_parameter(
-      Parameter<double>("A", 1.0), clproto::MessageType::PARAMETER_MESSAGE, clproto::ParameterMessageType::DOUBLE
-  );
-  test_encode_decode_parameter(
-      Parameter<std::vector<double>>("A", {1.0, 2.0, 3.0}), clproto::MessageType::PARAMETER_MESSAGE,
-      clproto::ParameterMessageType::DOUBLE_ARRAY
-  );
-  test_encode_decode_parameter(
-      Parameter<bool>("A", true), clproto::MessageType::PARAMETER_MESSAGE, clproto::ParameterMessageType::BOOL
-  );
-  test_encode_decode_parameter(
-      Parameter<std::vector<bool>>("A", {true, false, true, false}), clproto::MessageType::PARAMETER_MESSAGE,
-      clproto::ParameterMessageType::BOOL_ARRAY
-  );
-  test_encode_decode_parameter(
-      Parameter<std::string>("A", "value"), clproto::MessageType::PARAMETER_MESSAGE,
-      clproto::ParameterMessageType::STRING
-  );
-  test_encode_decode_parameter(
-      Parameter<std::vector<std::string>>("A", {"value 1", "value 2", "value 3"}),
-      clproto::MessageType::PARAMETER_MESSAGE, clproto::ParameterMessageType::STRING_ARRAY
-  );
-  Eigen::VectorXd vector_value(5);
-  vector_value << 1.0, 2.0, 3.0, 4.0, 5.0;
-  test_encode_decode_parameter(
-      Parameter<Eigen::VectorXd>("A", vector_value), clproto::MessageType::PARAMETER_MESSAGE,
-      clproto::ParameterMessageType::VECTOR
-  );
-  Eigen::MatrixXd matrix_value(2, 3);
-  matrix_value << 1.0, 2.0, 3.0, 4.0, 5.0, 6.0;
-  test_encode_decode_parameter(
-      Parameter<Eigen::MatrixXd>("A", matrix_value), clproto::MessageType::PARAMETER_MESSAGE,
-      clproto::ParameterMessageType::MATRIX
-  );
+TYPED_TEST_P(ParameterProtoTest, EncodeDecodeParameter) {
+  auto param = Parameter<TypeParam>("test", std::get<0>(this->test_case_));
+  clproto::test_encode_decode<Parameter<TypeParam>>(
+      param, clproto::PARAMETER_MESSAGE, test_parameter_equal<Parameter<TypeParam>>, std::get<1>(this->test_case_));
+  param.reset();
+  // FIXME(#50): Encoding of empty parameters is incorrect
```

Comment:
In order to make this work, we'll need another PR. I already have the fix but I wanted to create the test first and then make it work